### PR TITLE
Issue/2152 Item 3 OPA Policy Development

### LINKED ISSUE #2152: End to end basic "Approver" functionality: Item 3 OPA Policy Development
### This is the break-down task `item 3` for  #2140 

OPA policy development: translate this issue access control story into Rego policy. Mainly:
  - Some infrastructure code:
    - Decode JWT and get User ID (If failed to decode, assume it's anonymous users)
    - Retrieve current User info from `Auth API / whoami` API 
  - Cover access control of draft / non-draft datasets (as per ticket story)
  - Cover access control of header items (so that a new `Drafts` tab is only available to certain users. e.g. Approver)
    - We probably will leave this demo on a separate branch without merging it into master. However, if want to, it allows us can deploy the same version to DGA without impacting its functionality

> Item 3 Estimate: 8
> Blocked by item 1

### PR DESCRIPTION
**Pull request by t83714**
### What this PR does

Fixes #2152 
OPA Policy Development
- Cover access control of draft / non-draft datasets (as per ticket story)
- Cover access control of header items (so that a new Drafts tab is only available to certain users. e.g. Approver)
- There are more OPA policies changes on issue/2153

Commits after issue/2151:

https://github.com/magda-io/magda/compare/issue/2151...issue/2152

### Test

There are OPA test cases included for policies

To run it, install OPA (MACOS only. For other platforms, please use docker or check installation doc):
```bash
cd /usr/local/bin
curl -L -o opa https://github.com/open-policy-agent/opa/releases/download/v0.10.7/opa_darwin_amd64
chmod +x ./opa
```

To Run test cases locally:
```bash
cd deploy/helm/magda/charts/opa
opa test ./policies
```

You also can use docker to run test cases:
```bash
docker run -v $PWD/deploy/helm/magda/charts/opa/policies:/policies openpolicyagent/opa:0.10.7 test -v ./policies
```

### Merge instruction

As issue/2154 includes everything and more complete, we should merge that PR instead.
This PR created only for reviewing the history

### Checklist

-   [x] There are unit tests to verify my changes are correct
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+
+test_allow_no_constraints_permission_user_without_no_constraints_permission {
+    not allow with input as {
+        "operationUri": "object/dataset/draft/read",
+        "object": {
+            "dataset": {
+                "accessControl": {
+                    "ownerId": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx",
+                    "orgUnitOwnerId": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx",
+                    "preAuthoisedPermissionIds": ["xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx"]
```
Comment (AlexGilleran) [inline deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: Authorised?

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+                        "72d52505-cf96-47b2-9b74-d0fdc1f5aee7"
+                    ]
+                }
+            ],
+            "source": "ckan"
+        }
+    }
+}
+
+test_allow_pre_authoised_constraint_permission_user_permission_is_not_pre_authorised {
```
Comment (AlexGilleran) [inline deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: authorised

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+                        "72d52505-cf96-47b2-9b74-d0fdc1f5aee7"
+                    ]
+                }
+            ],
+            "source": "ckan"
+        }
+    }
+}
+
+test_allow_pre_authoised_constraint_permission_user_permission_is_pre_authorised {
```
Comment (AlexGilleran) [inline deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: authorised

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+                        "72d52505-cf96-47b2-9b74-d0fdc1f5aee7"
+                    ]
+                }
+            ],
+            "source": "ckan"
+        }
+    }
+}
+
+test_allow_pre_authoised_constraint_permission_user_permission_is_pre_authorised_for_incorrect_publishing_state {
```
Comment (AlexGilleran) [inline deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: authorised

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego (hunk 1 above)
Comment (t83714) [reply deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: You found it~ 😭 I just found it last week and it's corrected on issue/2154 with this commit:
https://github.com/magda-io/magda/pull/2175/commits/609798d0ce56d2e88d19415b557cf750d425d855

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego (hunk 2 above)
Comment (t83714) [reply deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: corrected on issue/2154 with this commit:
https://github.com/magda-io/magda/pull/2175/commits/609798d0ce56d2e88d19415b557cf750d425d855

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego (hunk 3 above)
Comment (t83714) [reply deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: corrected on issue/2154 with this commit:
https://github.com/magda-io/magda/pull/2175/commits/609798d0ce56d2e88d19415b557cf750d425d855

File: deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego (hunk 4 above)
Comment (t83714) [reply deploy/helm/magda/charts/opa/policies/object/dataset/allow_test.rego]: corrected on issue/2154 with this commit:
https://github.com/magda-io/magda/pull/2175/commits/609798d0ce56d2e88d19415b557cf750d425d855

**Comment by AlexGilleran** [comment]
Closing, changes made downstream.

**Pull request closed without merging**